Training/Training/harness.c

```diff
@@ -0,0 +1,127 @@
+      &si,             // Pointer to STARTUPINFO structure
+      &pi))            // Pointer to PROCESS_INFORMATION structure
+   {
+      // If CreateProcess fails, print an error message
+      int err = GetLastError ();
+      printf ("Failed to start process. Error: %lu\n", GetLastError ());
+      free (cmdline);
+      return 1;
+   }
+
+   // Wait until the process has finished execution
+   WaitForSingleObject (pi.hProcess, INFINITE);
+
+   // Close handles
+   CloseHandle (pi.hProcess);
+   CloseHandle (pi.hThread);
+
+   free (cmdline);
+   return 0;
+
+}
+
+/// <summary> Returns whether the given two files are equal or not </summary>
+int CompareFiles (char* file, char* tempOutPath, int* errBit, int* errBitValue) {
+   FILE* f1 = fopen (tempOutPath, "r"), * f2 = fopen (file, "r");
+   if (f1 == NULL || f2 == NULL) return -1;
+   int refFileSize = filelength (fileno (f1)) + 1,
+      outFileSize = filelength (fileno (f2)) + 1, j = 0, result = 1; // 1 means files are equal
+   char* refFileString = (char*)malloc (refFileSize * sizeof (char)),
+      * outFileString = (char*)malloc (outFileSize * sizeof (char));
+   fgets (refFileString, refFileSize, f1);
+   fgets (outFileString, outFileSize, f2);
+   char refFileChar = refFileString[j], outFileChar = outFileString[j];
+   while (refFileChar != '\0' || outFileChar != '\0') {
+      (*errBit)++;
+      if (refFileChar != outFileChar) {
+         result = 0; // 0 means files are not equal
+         *errBitValue = refFileChar - '0';
+         break;
+      }
+      refFileChar = refFileString[++j];
+      outFileChar = outFileString[j];
+   }
+   free (refFileString);
+   free (outFileString);
+   fclose (f1);
+   fclose (f2);
+   return result;
+}
+
+/// <summary>
+/// Test Harness
+/// </summary>
+/// <param name="argc"></param>
+/// <param name="argv">argv[1] is the name of the FSM</param>
+/// <returns></returns>
+int main (int argc, char** argv) {
+#define NTESTS 5                    // change this according to the number of tests you design. The more unique and relevant test patterns, the more complete your testing.
+   printf ("FSM Test Harness\n");
+   if (argc != 2) {
+      printf ("Usage: %s <FSM executable name>\n,", argv[0]);
+      return -1;
+   }
+   for (int i = 0; i < NTESTS; i++) {
+      char inpPath[MAX_PATH], outPath[MAX_PATH], * tempOutPath = "C:\\etc\\temp.txt";
+      sprintf (inpPath, "TData\\Input%d.txt", i + 1);
+      sprintf (outPath, "TData\\ExpOutput%d.txt", i + 1);
+      if (ExecProgram (argv[1], inpPath, tempOutPath) != 0)  // change the name of the input and output files in each set.
+         printf ("Error executing test %d\n", i + 1);
+      else {
+         int errBit = 0, errBitValue = 0, result = CompareFiles (outPath, tempOutPath, &errBit, &errBitValue), crtBitValue = errBitValue ? 0 : 1;
+         sprintf (inpPath, "Input%d.txt", i + 1);
+         if (result) printf ("No error testing %s\n", inpPath);
```

Comment:
a) How come you have "if else if else" testing a single conditional variable "result". Will the last else condition ever be executed?
b) You are using an integer variable like a boolean. This code is buggy.